src/v/net/connection.cc

```diff
@@ -12,8 +12,46 @@
 #include "net/exceptions.h"
 #include "rpc/service.h"
 
+#include <gnutls/gnutls.h>
+
 namespace net {
 
+/**
+ * Identify error cases that should be quickly retried, e.g.
+ * TCP disconnects, timeouts. Network errors may also show up
+ * indirectly as errors from the TLS layer.
+ */
+bool is_reconnect_error(const std::system_error& e) {
+    auto v = e.code().value();
+
+    // The name() of seastar's gnutls_error_category class
+    constexpr std::string_view gnutls_category_name{"GnuTLS"};
+
+    if (e.code().category().name() == gnutls_category_name) {
+        switch (v) {
+        case GNUTLS_E_PUSH_ERROR:
+        case GNUTLS_E_PULL_ERROR:
+        case GNUTLS_E_PREMATURE_TERMINATION:
+            return true;
+        default:
+            return false;
+        }
+    } else {
```

Comment:
None of this is type safe, in that the category isn't checked.

For example: `ECONNREFUSED` is equivalent to `std::errc::connection_refused`, `ETIMEDOUT` is `std::errc::timed_out` There may be several categories as is evidenced by `gnutls_category_name`.

It's probably safer to check the category here.

Reply:
oh you mean the else case in that we should be checking if it is system_category or something?

additionally we should probably try to expose gnutls_category form seastar instead of relying on the name string?